File: is_float/perl/is_float_flag.pl
use strict;
use warnings;

my ($plus, $minus, $point, $zero, $nine, $uc_e, $lc_e) = unpack("C*", "+-.09Ee");

sub is_float {
	my $s = shift;
	my $state = 0;
	for my $b (unpack("C*", $s)) {
		if ($state == 0) {
			if ($b == $plus || $b == $minus) {
				$state = 1;
			} elsif ($b == $point) {
				$state = 2;
			} elsif ($zero <= $b && $b <= $nine) {
				$state = 3;
			} else {
				$state = -1;
			}
		} elsif ($state == 1) {
			if ($b == $point) {
				$state = 2;
			} elsif ($zero <= $b && $b <= $nine) {
				$state = 3;
			} else {
				$state = -1;
			}
		} elsif ($state == 2) {
			if ($zero <= $b && $b <= $nine) {
				$state = 4;
			} else {
				$state = -1;
			}
		} elsif ($state == 3) {
			if ($b == $point) {
				$state = 4;
			} elsif ($zero <= $b && $b <= $nine) {
				$state = 3;
			} elsif ($b == $uc_e || $b == $lc_e) {
				$state = 5;
			} else {
				$state = -1;
			}
		} elsif ($state == 4) {
			if ($zero <= $b && $b <= $nine) {
				$state = 4;
			} elsif ($b == $uc_e || $b == $lc_e) {
				$state = 5;
			} else {
				$state = -1;
			}
		} elsif ($state == 5) {
			if ($b == $plus || $b == $minus) {
				$state = 6;
			} elsif ($zero <= $b && $b <= $nine) {
				$state = 7;
			} else {
				$state = -1;
			}
		} elsif ($state == 6) {
			if ($zero <= $b && $b <= $nine) {
				$state = 7;
			} else {
				$state = -1;
			}
		} elsif ($state == 7) {
			if ($zero <= $b && $b <= $nine) {
				$state = 7;
			} else {
				$state = -1;
			}
		}
		if ($state < 0) {
			next;
		}
	}

	return $state == 3 || $state == 4 || $state == 7;
}

my @tests = (
	["", 0],
	["+", 0],
	["-", 0],
	[".", 0],
	["+.", 0],
	["-.", 0],
	["+1+", 0],
	["-1-", 0],
	["+1", 1],
	["-1", 1],
	["1", 1],
	["+1.", 1],
	["-1.", 1],
	["1.", 1],
	["+.1", 1],
	["-.1", 1],
	[".1", 1],
	["+1.1", 1],
	["-1.1", 1],
	["1.1", 1],
	["+1.1E", 0],
	["+1.1E+", 0],
	["+1.1E-", 0],
	["+1.1E.", 0],
	["+1.1E1", 1],
	["+1.1E+1", 1],
	["+1.1E-1", 1],
	["+1.1E.1", 0],
	["+1.1E1.", 0],
	["+1.1e+1", 1],
	["+1234567890.1234567890E+1234567890", 1],
);

for my $t (@tests) {
	print "'$t->[0]'\t", $t->[1] == is_float($t->[0]) ? "OK" : "NG", "\n";
}

File: is_float/go/is_float_flag.go
package main

import "fmt"

func isFloat(s string) bool {
	state := 0
	for _, b := range []byte(s) {
		if state == 0 {
			if b == '+' || b == '-' {
				state = 1
			} else if b == '.' {
				state = 2
			} else if '0' <= b && b <= '9' {
				state = 3
			} else {
				state = -1
			}
		} else if state == 1 {
			if b == '.' {
				state = 2
			} else if '0' <= b && b <= '9' {
				state = 3
			} else {
				state = -1
			}
		} else if state == 2 {
			if '0' <= b && b <= '9' {
				state = 4
			} else {
				state = -1
			}
		} else if state == 3 {
			if b == '.' {
				state = 4
			} else if '0' <= b && b <= '9' {
				state = 3
			} else if b == 'E' || b == 'e' {
				state = 5
			} else {
				state = -1
			}
		} else if state == 4 {
			if '0' <= b && b <= '9' {
				state = 4
			} else if b == 'E' || b == 'e' {
				state = 5
			} else {
				state = -1
			}
		} else if state == 5 {
			if b == '+' || b == '-' {
				state = 6
			} else if '0' <= b && b <= '9' {
				state = 7
			} else {
				state = -1
			}
		} else if state == 6 {
			if '0' <= b && b <= '9' {
				state = 7
			} else {
				state = -1
			}
		} else if state == 7 {
			if '0' <= b && b <= '9' {
				state = 7
			} else {
				state = -1
			}
		}
		if state < 0 {
			break
		}
	}

	return state == 3 || state == 4 || state == 7
}

func main() {
	tests := []struct {
		s string
		b bool
	}{
		{"", false},
		{"+", false},
		{"-", false},
		{".", false},
		{"+.", false},
		{"-.", false},
		{"+1+", false},
		{"-1-", false},
		{"+1", true},
		{"-1", true},
		{"1", true},
		{"+1.", true},
		{"-1.", true},
		{"1.", true},
		{"+.1", true},
		{"-.1", true},
		{".1", true},
		{"+1.1", true},
		{"-1.1", true},
		{"1.1", true},
		{"+1.1E", false},
		{"+1.1E+", false},
		{"+1.1E-", false},
		{"+1.1E.", false},
		{"+1.1E1", true},
		{"+1.1E+1", true},
		{"+1.1E-1", true},
		{"+1.1E.1", false},
		{"+1.1E1.", false},
		{"+1.1e+1", true},
		{"+1234567890.1234567890E+1234567890", true},
	}
	for _, t := range tests {
		fmt.Printf("'%s'\t%s\n", t.s, map[bool]string{true: "OK", false: "NG"}[t.b == isFloat(t.s)])
	}
}
